zipline/assets/asset_writer.py

```diff
@@ -0,0 +1,744 @@
+    'first_traded',
+    'exchange',
+    'notice_date',
+    'root_symbol',
+    'expiration_date',
+    'contract_multiplier',
+    # The following fields are for compatibility with other systems
+    'file_name',  # Used as symbol
+    'company_name',  # Used as asset_name
+    'start_date_nano',  # Used as start_date
+    'end_date_nano'  # Used as end_date
+]
+
+# Expected fields for an Asset's metadata
+ASSET_TABLE_FIELDS = [
+    'sid',
+    'symbol',
+    'asset_name',
+    'start_date',
+    'end_date',
+    'first_traded',
+    'exchange'
+]
+
+# Expected fields for an Asset's metadata
+FUTURE_TABLE_FIELDS = ASSET_TABLE_FIELDS + [
+    'root_symbol_id',
+    'notice_date',
+    'expiration_date',
+    'contract_multiplier',
+]
+
+EQUITY_TABLE_FIELDS = ASSET_TABLE_FIELDS
+
+EXCHANGE_TABLE_FIELDS = [
+    'exchange_id',
+    'exchange',
+    'timezone'
+]
+
+ROOT_SYMBOL_TABLE_FIELDS = [
+    'root_symbol_id',
+    'root_symbol',
+    'sector',
+    'description',
+    'exchange_id'
+]
+
+
+class AssetDBWriter(with_metaclass(ABCMeta)):
+    """
+    Class used to write arbitrary data to SQLite database.
+    Concrete subclasses will implement the logic for a specific
+    input datatypes by implementing the gen_data method.
+
+    Methods
+    -------
+    write_all(db_conn, fuzzy_char=None, allow_sid_assignment=True,
+              constraints=False)
+        Write the data supplied at initialization to the database.
+    init_db(db_conn, constraints=False)
+        Create the SQLite tables (called by write_all).
+    gen_data(self, equities, futures, exchanges, root_symbols)
+        Returns data in standard format.
+
+    """
+
+    def write_all(self, db_conn, fuzzy_char=None, allow_sid_assignment=True,
+                  constraints=False):
+        """ Write pre-supplied data to SQLite.
+
+        Parameters
+        ----------
+        db_conn: sqlite3.Connection
+            A connection to a SQLite database.
+        fuzzy_char: string
+            A string for use in fuzzy matching.
+        allow_sid_assignment: boolean
+            If True then the class can assign sids where necessary.
+        constraints: boolean
+            If True, create SQL ForeignKey and Index constraints.
+
+        """
+
+        self.allow_sid_assignment = allow_sid_assignment
+        if allow_sid_assignment:
+            ts = normalize_date(pd.Timestamp('now', tz='UTC'))
+            # Store as seconds since UNIX Epoch for compatibility
+            # with SQL.
+            self.end_date_to_assign = (ts.value // 10 ** 9)
+
+        # Create SQL tables
+        self.init_db(db_conn, constraints)
+
+        # Get the data to add to SQL
+        equities, futures, exchanges, root_symbols = self.gen_data()
+
+        c = db_conn.cursor()
+        # Write to the SQL tables, using the raw SQL driver instead
+        # of the pandas.DataFrame.to_sql method, as the former
+        # allows us to create an SQL transaction
+        c.execute('BEGIN')
```

Comment:
This should just be using a the cursor as a context manager:

```
with c:
    # do stuff
```
This will commit if everything in the context succeeds, and will rollback if an exception is raised.